### PR TITLE
Implement reporting unit test timings in CI

### PR DESCRIPTION
**Pull request by meddle0x53**
## Motivation

Implements https://github.com/ProvableHQ/snarkVM/issues/2981 and also stores the timing reports as CSVs so they can be analysed.

## Test Plan

The CI passes and the timings of the different tests could be viewed.

## Documentation

The PR updates the CI configuration and modifies how tests are ran.

## Backwards compatibility

Backwards compatible.

### CODE REVIEW

**Comment by vicsn** [comment]
Had a chat with Lukasz, nightly does have a higher chance of inadvertently breaking things, could you use `cargo nextest` instead? That also has an option to report timings, but it may be a bit more work to update the circleci config.

Just for context: the reason we use limited test-threads is to prevent going out of memory.

File: .circleci/config.yml [hunk 1]
```diff
@@ -149,14 +149,136 @@ commands:
...
           name: "Run Tests"
           timeout: << parameters.timeout >>
           no_output_timeout: << parameters.no_output_timeout >>
-          command: RUST_MIN_STACK=67108864 cargo test --package=<< parameters.workspace_member >> << parameters.flags >>
+          command: |
+            set -euo pipefail
+            export CARGO_TERM_COLOR=never
+            mkdir -p artifacts target/nextest/ci
+
+            OUT="artifacts/<< parameters.workspace_member >>-nextest.txt"
```
Comment (meddle0x53) [inline .circleci/config.yml:170]: The new way of testing measures time and stores the results as artifacts:
1. The `*nextest.txt` contains the test logs
2. The `-test-times.csv` contains all the tests with their times in a CSV table
3. The `-slow-tests-top30.csv` contains the 30 slowest sorted from slowest downwards.

File: .circleci/config.yml [hunk 2]
```diff
@@ -149,14 +149,136 @@ commands:
...
+            CSV="artifacts/<< parameters.workspace_member >>-test-times.csv"
+            TOP="artifacts/<< parameters.workspace_member >>-slow-tests-top30.csv"
+            JUNIT="target/nextest/ci/junit.xml"
+
+            FLAGS_FULL="<< parameters.flags >>"
```
Comment (meddle0x53) [inline .circleci/config.yml:176]: The flags passed to nextest and the flags passed to the different cases shouldn't interfere so with a lot of testing I reworked this script to work with all possible types of flags being passed when ran with `cargo nextest` and its flags.

File: .circleci/config.yml [hunk 3]
```diff
@@ -149,14 +149,136 @@ commands:
...
+            fi
+
+            BUILD_FLAGS_SANITIZED="$(printf '%s' "$BUILD_FLAGS" \
+              | sed -E "s/(^|[[:space:]])--test[[:space:]]+'[^']*'([[:space:]]|$)/ /g" \
+              | sed -E 's/(^|[[:space:]])--test[[:space:]]+\"[^\"]*\"([[:space:]]|$)/ /g' \
+              | sed -E 's/(^|[[:space:]])--test[[:space:]]+[^[:space:]]+([[:space:]]|$)/ /g' \
+              | tr -s ' ' | sed -E 's/^[[:space:]]+//; s/[[:space:]]+$//' \
+            )"
+
+            FILTER_EXPR=""
```
Comment (meddle0x53) [inline .circleci/config.yml:206]: The nextest syntax for filtering is a bit strange, so I did this here to work with the old test flags for filtering.

So when defining a new package tests we can just use the normal `--skip` flags, this little script handles converting them.

File: .circleci/config.yml [hunk 4]
```diff
@@ -149,14 +149,136 @@ commands:
...
+                  fi
+                done < "$tmpf"
+                rm -f "$tmpf"
+                if [ -n "$EXPRS" ]; then
+                  FILTER_EXPR="not ($EXPRS)"
+                fi
+              fi
+            fi
+
+            JOBS="${NEXTEST_JOBS:-2}"
```
Comment (meddle0x53) [inline .circleci/config.yml]: Next test's number of JOBS are defined here and can be overridden by passing `environment` variables to the specific package test definitions. By default it is the minimal 2, but for some I adjusted more.

File: .circleci/config.yml [hunk 5]
```diff
@@ -149,14 +149,136 @@ commands:
...
+              CFG_ARGS+=(--config-file ".config/nextest.toml")
+            fi
+            {
+              printf '%s\n' '[profile.ci]'
+              printf '%s\n' 'fail-fast = false'
+              printf '%s\n' "slow-timeout = \"${SLOW_PERIOD}\""
+              printf '%s\n' ''
+              printf '%s\n' '[profile.ci.junit]'
+              printf '%s\n' 'path = "junit.xml"'
+            } > "$TMPD/nextest-override.toml"
```
Comment (meddle0x53) [inline .circleci/config.yml:273]: Nextest needs configuration file, so this is creating it.

File: .circleci/config.yml [hunk 6]
```diff
@@ -149,14 +149,136 @@ commands:
...
+            CFG_ARGS+=(--config-file "$TMPD/nextest-override.toml")
+
+            echo "==> nextest package: << parameters.workspace_member >>"
+            echo "==> build flags (sanitized): '${BUILD_FLAGS_SANITIZED}'"
+            echo "==> filter-expr: ${FILTER_EXPR:-<none>}  (right-of-sep: ${RIGHT_OF_SEP:-<none>})"
+            echo "==> jobs: ${JOBS}"
+
+            if printf '%s' " $BUILD_FLAGS_SANITIZED " | grep -q ' --no-run '; then
```
Comment (meddle0x53) [inline .circleci/config.yml:285]: `--no-run` breaks nextest, so if we are going to do a not run test, we just keep the old command.

File: .circleci/config.yml [hunk 7]
```diff
@@ -149,14 +149,136 @@ commands:
...
+              cargo test --package="<< parameters.workspace_member >>" $BUILD_FLAGS_SANITIZED | tee "$OUT"
+              : > "$CSV"; : > "$TOP"
+            else
+              cargo nextest run \
```
Comment (meddle0x53) [inline .circleci/config.yml]: Here is the nextest comand working and fully tested.

File: .circleci/config.yml [hunk 8]
```diff
@@ -149,14 +149,136 @@ commands:
...
+                -p "<< parameters.workspace_member >>" $BUILD_FLAGS_SANITIZED \
+                --profile ci -j "${JOBS}" "${CFG_ARGS[@]}" \
+                --status-level fail --hide-progress-bar \
+                --no-tests=pass \
+                ${FILTER_EXPR:+--filter-expr "$FILTER_EXPR"} \
+                2>&1 | tee "$OUT"
+
+              if [ -f "$JUNIT" ]; then
+                awk -v RS='<testcase ' '
```
Comment (meddle0x53) [inline .circleci/config.yml:299]: This is the generation of the artifacts from the test logs. In nextest timing is reported in `<time>`, so we `awk` it and the name of the test.

File: .circleci/config.yml [hunk 9]
```diff
@@ -149,14 +149,136 @@ commands:
...
+                  NR>1 {
+                    name=""; classname=""; time="";
+                    start=index($0,"classname=\""); if(start>0){ start+=11; rest=substr($0,start); end=index(rest,"\""); if(end>0) classname=substr(rest,1,end-1) }
+                    start=index($0,"name=\"");      if(start>0){ start+=6;  rest=substr($0,start); end=index(rest,"\""); if(end>0) name=substr(rest,1,end-1) }
+                    start=index($0,"time=\"");      if(start>0){ start+=6;  rest=substr($0,start); end=index(rest,"\""); if(end>0) time=substr(rest,1,end-1) }
+                    if(name!="" && time!=""){ full=(classname!=""?classname"::"name:name); printf "%s,%s\n", full, time }
+                  }
+                ' "$JUNIT" > "$CSV" || true
+
+                sort -t, -k2,2nr "$CSV" | head -n 30 > "$TOP" || true
```
Comment (meddle0x53) [inline .circleci/config.yml:310]: Here is how the top 30 are sorted and created.

File: .circleci/config.yml [hunk 10]
```diff
@@ -149,14 +149,136 @@ commands:
...
+              else
+                echo "WARN: $JUNIT not found; no timings extracted." >&2
+                : > "$CSV"; : > "$TOP"
+              fi
+            fi
+
+      - store_artifacts:
```
Comment (meddle0x53) [inline .circleci/config.yml:317]: This stores the new artifacts  and they can be viewed in the CI job.

File: .circleci/config.yml [hunk 11]
```diff
@@ -149,14 +149,136 @@ commands:
...
+
+            OUT="artifacts/<< parameters.workspace_member >>-nextest.txt"
+            CSV="artifacts/<< parameters.workspace_member >>-test-times.csv"
+            TOP="artifacts/<< parameters.workspace_member >>-slow-tests-top30.csv"
+            JUNIT="target/nextest/ci/junit.xml"
+
+            FLAGS_FULL="<< parameters.flags >>"
+            FLAGS_TRIM="$(printf '%s' "$FLAGS_FULL" | sed -e 's/^[[:space:]]*//' -e 's/[[:space:]]*$//')"
+
+            RIGHT_OF_SEP=""
```
Comment (vicsn) [inline .circleci/config.yml:180]: Can you add a comment for each block what it does and why it is needed?

File: .circleci/config.yml [hunk 12]
```diff
@@ -601,11 +728,14 @@ jobs:
...
           flags: --lib --bins
-          timeout: 20m
+          no_output_timeout: 20m
+          timeout: 30m
 
   synthesizer-test:
     executor: rust-docker
     resource_class: << pipeline.parameters.twoxlarge >>
+    environment:
+      NEXTEST_JOBS: "4"
```
Comment (vicsn) [inline .circleci/config.yml]: Does this replace `--test-threads=2`? Should we replace all instances of `--test-threads` by `NEXTEST_JOBS`?

Perhaps related: the synthesizer-workflow now takes 55 minutes, I recall it taking less in previous runs.

**Comment by vicsn** [comment]
Looks great! Can you make a PR in snarkOS as well?

File: .circleci/config.yml (hunk 11 above)
Comment (meddle0x53) [reply .circleci/config.yml:180]: I didn't want to add bash comments to the inlined script to not break something unintentionally in the logic, but will try!

File: .circleci/config.yml (hunk 12 above)
Comment (meddle0x53) [reply .circleci/config.yml]: Ugh, yeah, so I'll try to make `--test-threads` be automatically converted to `NEXTEST_JOBS` so we don't have to change anything.

For the workflow - I think it may be from the cache, being invalidated with each of my changes or maybe related to not using the `--test-threads`. Let me do a commit to see.

**Comment by meddle0x53** [comment]
@vicsn I tried a few things here (though am mostly trying to make the snarkos version work in the moment) and it seems nextest runs are slower.

So, because we are going to try to speed up things anyway either I can try to make `synthesizer-workflow` faster in this PR or in a next one - your decision.

File: .circleci/config.yml (hunk 12 above)
Comment (vicsn) [reply .circleci/config.yml]: > @vicsn I tried a few things here (though am mostly trying to make the snarkos version work in the moment) and it seems nextest runs are slower.
> So, because we are going to try to speed up things anyway either I can try to make synthesizer-workflow faster in this PR or in a next one - your decision.

Let's not merge when it's twice as slow. And independent of the runtime, it would be great if we could consistently use `NEXTEST_JOBS` OR `--test-threads`, if those are conceptually the same thing

**Comment by meddle0x53** [comment]
Actually looking around many workflows are passing for the same time, especially in the SnarkOS version. It is specific for the [synthesizer-workflow](https://github.com/ProvableHQ/snarkVM/pull/2995/checks?check_run_id=54468695960) and for some of the others a bit, but not so much, hmmm.

File: .circleci/config.yml (hunk 12 above)
Comment (meddle0x53) [reply .circleci/config.yml]: I think I found the problem, debugging, checking and I may be able to fix it `--test-threads ` will be used only.

**Review by vicsn — APPROVED** [review]
Amazing!

**Comment by vicsn** [comment]
We will surely hit some edge-cases the following days on new PRs, but happy that we have this data to start optimizing
